src/lib/operators/sort.cpp

```diff
@@ -1,260 +1,283 @@
+        const auto chunk = _table_in->get_chunk(chunk_id);
+        Assert(chunk, "Did not expect deleted chunk here.");  // see https://github.com/hyrise/hyrise/issues/1686
 
-      auto base_segment = chunk->get_segment(_column_id);
+        auto base_segment = chunk->get_segment(_column_id);
 
-      segment_iterate<SortColumnType>(*base_segment, [&](const auto& position) {
-        if (position.is_null()) {
-          null_value_rows.emplace_back(RowID{chunk_id, position.chunk_offset()}, SortColumnType{});
-        } else {
-          row_id_value_vector.emplace_back(RowID{chunk_id, position.chunk_offset()}, position.value());
-        }
-      });
+        segment_iterate<SortColumnType>(*base_segment, [&](const auto& position) {
+          if (position.is_null()) {
+            _null_value_rows.emplace_back(RowID{chunk_id, position.chunk_offset()}, SortColumnType{});
+          } else {
+            _row_id_value_vector.emplace_back(RowID{chunk_id, position.chunk_offset()}, position.value());
+          }
+        });
+      }
     }
   }
 
-  template <typename Comparator>
-  void _sort_with_operator() {
-    Comparator comparator;
-    std::stable_sort(_row_id_value_vector->begin(), _row_id_value_vector->end(),
-                     [comparator](RowIDValuePair a, RowIDValuePair b) { return comparator(a.second, b.second); });
+  // When there was a preceding sorting run, we materialize by retaining the order of the values in the passed PosList.
+  void _materialize_column_from_pos_list(const PosList& pos_list) {
+    const auto input_chunk_count = _table_in->chunk_count();
+    auto accessor_by_chunk_id =
+        std::vector<std::unique_ptr<AbstractSegmentAccessor<SortColumnType>>>(input_chunk_count);
+    for (auto input_chunk_id = ChunkID{0}; input_chunk_id < input_chunk_count; ++input_chunk_id) {
+      const auto& base_segment = _table_in->get_chunk(input_chunk_id)->get_segment(_column_id);
+      accessor_by_chunk_id[input_chunk_id] = create_segment_accessor<SortColumnType>(base_segment);
+    }
+
+    for (auto row_id : pos_list) {
+      const auto [chunk_id, chunk_offset] = row_id;
+
+      auto& accessor = accessor_by_chunk_id[chunk_id];
+      const auto typed_value = accessor->access(chunk_offset);
+      if (!typed_value) {
+        _null_value_rows.emplace_back(row_id, SortColumnType{});
+      } else {
+        _row_id_value_vector.emplace_back(row_id, typed_value.value());
+      }
+    }
   }
 
   const std::shared_ptr<const Table> _table_in;
 
   // column to sort by
   const ColumnID _column_id;
   const OrderByMode _order_by_mode;
-  // chunk size of the materialized output
-  const size_t _output_chunk_size;
 
-  std::shared_ptr<std::vector<RowIDValuePair>> _row_id_value_vector;
-  std::shared_ptr<std::vector<RowIDValuePair>> _null_value_rows;
+  std::vector<RowIDValuePair> _row_id_value_vector;
+
+  // Stored as RowIDValuePair even if value is unused for better type compatibility.
```

Comment:
Shouldn't that be "Stored as RowIDValuePair for better type compatibility even if value is unused."?